### DISCUSSION TITLE
Pseudonimization of report data

### DISCUSSION THREAD

**Discussion by rawc0der** [kubescape/kubescape#1201] [Ideas]
Hi @rakshitgondwal - lets discuss the details here.

Problem:
Reports contain sensitive information like Pod names, Namespaces, Resource names etc that are somerimes classified as strictly confidential to a specific group or even more restrictive individual person (C3, C4 data classification). 
For example the User has a Namespace with his name and a Pod name with a newly developed unreleased Product -- this is considered Confidential info, not supposed to be visible to anyone else.

The idea behind solving this is to use
1) a mapping file that can be encrypted with Secret key (pgp, SOPS, etc) of realName -> pseudoName mappings (friendly hunan readable dictionary words + uuid), which can be used to produce a more anonimized report which is saved up in the cloud or distributed to SecOps teams, etc.. 
2) process behind report conversion (enc/dec)
3) display layer

Let me know what you think about this whole idea, thanks!

**Comment by rakshitgondwal** [comment]
Hey @rawc0der,
This idea sounds good to me.
Correct me if I'm wrong here: A user would be able to hide sensitive information from his reports using a command like `kubescape scan --hide <field>` or something similar. The field's original value would be replaced with a psuedoName and the original value would be stored in a mapping file which would be encrypted with a secret key. The user would be able to later decipher the masking file using the secret key. Does this sounds good?

**Comment by rawc0der** [comment]
Yes, exactly - as a user you can use an optional CLI flag to transform a pure report to -> pseudonymised report. This involves integrating with some form of Cryptographic service for owning the Key for **Mapping File Encryption/Decryption** (this is the whole crux).

**Problem Scope**
I think this topic is mostly valuable in contexts of tightly regulated environments (more specifically when exchanging data with other vendors i.e. hosted environments) where enhanced data protection layers can be added for all kinds of information (PII: personal data, Metadata about internal domains: machine metrics data, etc)  

To add extra protection a Private Key from any external  Key Management Service can be used to Obfuscate/Pseudonymise entire report keys metadata `(all Kubernetes Resource names)` by working with a **Mapping File** encrypted by _Privately owned_ **KMS Provider** & **Key ID flags** combinations `--kms-provider  --kms-key`.

Possible CLI flags:
```
   --kms-provider       list of KMS Provider options (Azure Key Vault, GCP KMS, AWS KMS)
   --kms-key          resource IDs
```

Definitely recommend checking out [SOPS](https://github.com/mozilla/sops) for possible initial solution path. 

**Desired outcome (least technical effort):**
Compared to masking all data by default, let's also thing about including some extra choices: 
- ``--filter-keys-regex`` can be a optional feature of using regex selection on JSON keys 
- and ``no flag`` the by default is doing obfuscation/anonymisation on entire Kubernetes resource key names (while maybe keeping the data intact for statistics and DevSecOps reports) 🚒 

What do you think?

**Comment by rakshitgondwal** [comment]
This sounds good to me but I just had some questions in my mind.
- Are you suggesting to Pseudonymise the whole Kubernetes resource names or some specific fields? or are you suggesting having both?
- Would the key provided in `--kms-key` be the same key that would be used to encrypt the mapping file?
- Where do you think should we store the mapping file?

**Reply by rawc0der** [reply]
Good questions!

1. Are you suggesting to Pseudonymise the whole Kubernetes resource names or some specific fields? or are you suggesting having both?
- Well, if I think about it - having some sane defaults is probably more desirable (maybe some kind of `preset jsonPath expressions` => `.items[*]['metadata.name']`, `.items[*]['metadata.namespace']`, `.items[*]['metadata.labels']`, Still figuring out impacted surface of PII data in reports etc ).
- But if you think about how simple it would be to provide some defaults and allow to override:

```bash
# using --hide parameter with global defaults (fallback to global .pseudo configuration file for hidden field rules)
kubescape scan --hide

# using --hide parameter with inline value (should be exclusive with the global config file)
kubescape scan --hide ".items[*]['metadata.labels'], .items[*]['metadata.name']"
"
```

2. Would the key provided in --kms-key be the same key that would be used to encrypt the mapping file?
- Yes, that is the key I am referring to. Mapping file should only be deciphered by private key Owner. 
I think in this case the main problem is how can we integrate a 3'rd party tool in kubescape without adding too much noise to core components (referring to using External KMS Key from different vendors), maybe through some plugin system which is provided for extending kubescape and allows client-side crypto operations using SOPS/ or other crypto kit? cc @slashben?

3. Where do you think should we store the mapping file?
- I think initially it can be local output if user doesen't want **report data** persisted, and if development work goes smooth on future iterations of this feature (🤞 😄 ),  the **Pseudonimized report data**  can be also safely shared with **ARMO Cloud** (when using `--submit`) - _this end result is actually what would be most interesting_. 

I know this thread and discussions are still in "embryonic phase", but hope the idea is starting to come together.

**Comment by rakshitgondwal** [comment]
Thank you for the clarification @rawc0der!
I have prepared a little diagram of how things might work. Please have a look.
![kubescape-hide](https://user-images.githubusercontent.com/98955085/235142294-dffc233c-e24b-479f-9c75-2ab388267975.png)

**Reply by rawc0der** [reply]
Looks good! 

Maybe we could have some kind of wrapper which prompts for the same parameters from sops? 

Also sops has a global config file as defaults, in case the user does not override with CLI arguments... you could consider having some default selector filters (for whatever the syntax would be, but do consider [JSONPath](https://kubernetes.io/docs/reference/kubectl/jsonpath/) as it's familiar to most `kubectl` users) in a config file. 
 
I'm thinking if this entire interface functionality can be contained under the same command group from `--hide`, then maybe this could be designed as a plugin "on top" of kubescape.

**Comment by shobhit9957** [comment]
Great Discussions!, Can I contribute to this too... like some sort of non-code contribution. Just a beginner, would love to be a part of it.

**Comment by jijo-OO7** [comment]
Hi @rawc0der i have planned  a phased approach and keep each changes small and reviewable.

**1.  (kubescape, minimal + self-contained):**

* Add an opt-in `--anonymize` flag wired into `ScanInfo`.
* Introduce a small `core/pkg/anonymizer` package.
* Hook anonymization right after `resultsHandling.SetData(scanData)` in `core/core/scan.go`, so it runs as a post-processing step.
* In the anonymizer:

  * Traverse `ScanData.AllResources` and replace `metadata.name` and `metadata.namespace` using `workloadinterface.InspectMap` / `SetInMap`.
  * Replace `ImageScanData.Image` values.
  * Build a simple in-memory mapping (`real -> pseudo`) and write it to `mapping.json` in the working directory.
* Keep this phase intentionally simple:

  * No config / JSONPath
  * No encryption
  * No changes to printers or OPA processing

**2.  (stability / determinism):**

* Make pseudonyms deterministic (hash-based, optional salt via `--anonymize-salt`).
* Ensure consistent mapping across runs given the same inputs.

**3. (field selection):**

* Add `--anonymize-fields` (start with a simple list, not full JSONPath initially).
* Default to `metadata.name`, `metadata.namespace`, and image names; allow users to override/extend.

**4.  (round-trip support):**

* Add a `kubescape deanonymize` command that uses the generated mapping to restore original values.
* Ensure the transformation is lossless for the fields we modify.

**5.  (mapping protection):**

* Add optional encryption for the mapping file (passphrase first).
* Consider KMS/SOPS integration as a follow-up once the core flow is stable.

I’ll start with Phase 1 and open a focused PR for that (flag + post-processing hook + minimal anonymizer + mapping output), then iterate in follow-ups.

I’d also really value your input on the `anonymizer` design before I open the PR specifically around the package shape and the exact fields we should be targeting in `OPASessionObj` (e.g., `metadata.name`, `metadata.namespace`, image references).

Are there other areas or data surfaces you think we should consider from the start, or any existing patterns in the codebase you’d prefer this to align with?
